File: src/two_means/test/test_diptest.R
## create share object using the following code in your bash
##  R CMD SHLIB -c test_diptest.c diptest.c dip.c utils.c

dyn.load("test_diptest.so")

## test euc_dists
set.seed(100)
X <- as.matrix(iris[sample(1:nrow(iris), 15), -5])
row.names(X) <- NULL
n <- nrow(X)
p <- ncol(X)
D <- as.matrix(dist(X))

ret <- .C("test_euc_dists",
          X = as.double(t(X)),
          n = as.integer(n),
          p = as.integer(p),
          D = double(0.5*n*(n-1))
          )

ret$D - D[upper.tri(D)]

## test diptst

x <- sort(rnorm(20))

diptest::dip(x)

ret <- .C("test_diptst",
          d = as.double(x),
          n = as.integer(length(x)),
          dip = double(1))


## test unimodarity test

seed <- 124
set.seed(seed)
Q <- MixSim::MixSim(0.01, p=2, K=2, sph=T, hom=T)
A <- MixSim::simdataset(1000, Pi=Q$Pi, Mu=Q$Mu, S=Q$S)
X <- A$X
n <- nrow(X)
p <- ncol(X)

D <- as.matrix(dist(X))

B <- 1000
dip_b <- rep(0, B)

set.seed(seed)
for (i in 1:B) dip_b[i] <- diptest::dip(runif(n-1))

dips <- rep(0, n)
for (i in 1:n) {
    d <- D[i, -i]
    dips[i] <- diptest::dip(d)
}

pvals <- rep(0, n)
for (i in 1:n)
    pvals[i] <- sum(dips[i] < dip_b) / B

print(sum(pvals < 1e-16) / n)

set.seed(seed)
ret <- .C("test_testCluster_unimodarity",
          X = as.double(t(X)),
          n = as.integer(n),
          p = as.integer(p),
          c = as.integer(0:(n-1)),
          a = as.double(1e-16),
          D = double(0.5*n*(n-1)))

File: src/two_means/test/test_diptest.c
#include "utils.h"
#include "dip.h"
#include "diptest.h"

void test_euc_dists(double *X, int *pn, int *pp, double *D) {
  int n = *pn, p = *pp;
  int i, j;
  double d;

  euc_dists(X, n, p, D);

  i = (int)(0.5 * n);

  printf("i=%d:\n", i);
  for (j = 0; j < i; j++) {
    d = D[(int)(0.5 * i * (i-1)) + j];
    printf(" %.4f", d);
  }
  for (j = i+1; j < n; j++) {
    d = D[(int)(0.5 * j * (j-1)) + i];
    printf( "%.4f ", d);
  } //d is distances from viewer ck[i]

  printf("\n");


}

void test_diptst(double *d, int *pn, double *dip) {
  int n = *pn;

  int low_hi[4];
  int ifault;
  int *gcm = (int *)malloc(sizeof(int) * (n));
  int *lcm = (int *)malloc(sizeof(int) * (n));
  int *mn = (int *)malloc(sizeof(int) * (n));
  int *mj = (int *)malloc(sizeof(int) * (n));
  const int min_is_0 = 0;
  const int debug_diptst = 0;

  diptst(d, pn, dip, low_hi, &ifault, gcm, lcm, mn, mj, &min_is_0, &debug_diptst);

  free(gcm); free(lcm); free(mn); free(mj);
}

void test_testCluster_unimodarity(double *X, int *pn, int *pp, int *c, double *pa, double *D) {
  int n = *pn, p = *pp;
  double a = *pa;

  euc_dists(X, n, p, D);

  testCluster_unimodality(D, n, c, a, 0.01, 1);
}
